src/NSubstitute/Core/SubstitutionContext.cs

```diff
@@ -14,123 +12,118 @@ public class SubstitutionContext : ISubstitutionContext
+        public SequenceNumberGenerator SequenceNumberGenerator { get; }
+        public IThreadLocalContext ThreadContext { get; }
 
         static SubstitutionContext()
         {
             Current = new SubstitutionContext();
         }
 
-        SubstitutionContext()
+        private SubstitutionContext()
         {
-            var callRouterFactory = new CallRouterFactory();
-            var argSpecificationQueue = new ArgumentSpecificationDequeue(DequeueAllArgumentSpecifications);
+            ThreadContext = new ThreadLocalContext();
+            SequenceNumberGenerator = new SequenceNumberGenerator();
+            _callRouterResolver = new CallRouterResolver();
+            RouteFactory = new RouteFactory(ThreadContext);
+
+            var callRouterFactory = new CallRouterFactory(SequenceNumberGenerator, RouteFactory);
+            var argSpecificationQueue = new ArgumentSpecificationDequeue(ThreadContext.DequeueAllArgumentSpecifications);
             var dynamicProxyFactory = new CastleDynamicProxyFactory(argSpecificationQueue);
             var delegateFactory = new DelegateProxyFactory(dynamicProxyFactory);
             var proxyFactory = new ProxyFactory(delegateFactory, dynamicProxyFactory);
-            var callRouteResolver = new CallRouterResolver();
-            _substituteFactory = new SubstituteFactory(this, callRouterFactory, proxyFactory, callRouteResolver);
+            SubstituteFactory = new SubstituteFactory(ThreadContext, callRouterFactory, proxyFactory, RouteFactory);
         }
 
-        public SubstitutionContext(ISubstituteFactory substituteFactory)
+        public SubstitutionContext(ISubstituteFactory substituteFactory,
+            IRouteFactory routeFactory,
+            IThreadLocalContext threadLocalContext,
+            ICallRouterResolver callRouterResolver)
         {
-            _substituteFactory = substituteFactory;
+            SubstituteFactory = substituteFactory ?? throw new ArgumentNullException(nameof(substituteFactory));
+            RouteFactory = routeFactory ?? throw new ArgumentNullException(nameof(routeFactory));
+            ThreadContext = threadLocalContext ?? throw new ArgumentNullException(nameof(threadLocalContext));
+            _callRouterResolver = callRouterResolver ?? throw new ArgumentNullException(nameof(callRouterResolver));
+
+            SequenceNumberGenerator = new SequenceNumberGenerator();
         }
 
-        public ISubstituteFactory SubstituteFactory { get { return _substituteFactory; } }
-        public SequenceNumberGenerator SequenceNumberGenerator { get { return _sequenceNumberGenerator; } }
-        public bool IsQuerying { get { return _currentQuery.Value != null; } }
+        public ICallRouter GetCallRouterFor(object substitute) =>
+            _callRouterResolver.ResolveFor(substitute);
+
+
+        // ***********************************************************
+        // ********************** OBSOLETE API **********************
```

Comment:
Should these members be marked `[Obsolete]` then?

Reply:
@dtchepak Nice catch, thanks! 👍 Due to some reasons I missed that I should mark implemented members as well :-/